federation/pkg/federation-controller/ingress/ingress-controller.go

```diff
@@ -0,0 +1,301 @@
+	extensions_v1beta1 "k8s.io/kubernetes/pkg/apis/extensions/v1beta1"
+	"k8s.io/kubernetes/pkg/client/cache"
+	"k8s.io/kubernetes/pkg/controller"
+	"k8s.io/kubernetes/pkg/controller/framework"
+	pkg_runtime "k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"github.com/golang/glog"
+)
+
+const (
+	IngressReviewDelay    = time.Second * 10
+	ClusterAvailableDelay = time.Second * 20
+	SmallDelay            = time.Second * 3
+	UpdateTimeout         = time.Second * 30
+
+	allClustersKey = "ALL_CLUSTERS"
+)
+
+type IngressController struct {
+	// For triggering single ingress reconcilation. This is used when there is an
+	// add/update/delete operation on a ingress in either federated API server or
+	// in some member of the federation.
+	ingressDeliverer *util.DelayingDeliverer
+
+	// For triggering all ingresss reconcilation. This is used when
+	// a new cluster becomes available.
+	clusterDeliverer *util.DelayingDeliverer
+
+	// Contains ingresss present in members of federation.
+	ingressFederatedInformer util.FederatedInformer
+	// For updating members of federation.
+	federatedUpdater util.FederatedUpdater
+	// Definitions of ingresss that should be federated.
+	ingressInformerStore cache.Store
+	// Informer controller for ingresss that should be federated.
+	ingressInformerController framework.ControllerInterface
+
+	// Client to federated api server.
+	federatedApiClient federation_release_1_4.Interface
+
+	stopChan chan struct{}
+}
+
+// A structure passed by delying deliver. It contains a ingress that should be reconciled and
+// the number of trials that were made previously and ended up in some kind of ingress-related
+// error (like failure to create).
+type ingressItem struct {
+	ingress string
+	trial   int64
+}
+
+// NewIngressController returns a new ingress controller
+func NewIngressController(client federation_release_1_4.Interface) *IngressController {
+	ic := &IngressController{
+		federatedApiClient: client,
+		stopChan:           make(chan struct{}),
+	}
+
+	// Build delivereres for triggering reconcilations.
+	ic.ingressDeliverer = util.NewDelayingDeliverer()
+	ic.clusterDeliverer = util.NewDelayingDeliverer()
+
+	// Start informer in federated API servers on ingresss that should be federated.
+	ic.ingressInformerStore, ic.ingressInformerController = framework.NewInformer(
+		&cache.ListWatch{
+			ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+				return client.Extensions().Ingresses("<all-namespaces>").List(options)
+			},
+			WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+				return client.Extensions().Ingresses("<all-namespaces>").Watch(options)
+			},
+		},
+		&extensions.Ingress{},
+		controller.NoResyncPeriodFunc(),
+		util.NewTriggerOnAllChanges(func(obj pkg_runtime.Object) { ic.deliverIngressObj(obj, 0, 0) }))
+
+	// Federated informer on ingresss in members of federation.
+	ic.ingressFederatedInformer = util.NewFederatedInformer(
```

Comment:
The user agent name of federatedInformer is `userAgentName     = "federation-service-controller"`, we need change this as a more comment name.